fix getStateRandao not returning historic RANDAO mix values

Muzry opened the pull request:
<!-- Thanks for sending a PR! Before submitting:

1. If this is your first PR, check out our contribution guide here https://docs.prylabs.network/docs/contribute/contribution-guidelines
   You will then need to sign our Contributor License Agreement (CLA), which will show up as a comment from a bot in this pull request after you open it. We cannot review code without a signed CLA.
2. Please file an associated tracking issue if this pull request is non-trivial and requires context for our team to understand. All features and most bug fixes should have
   an associated issue with a design discussed and decided upon. Small bug
   fixes and documentation improvements don't need issues.
3. New features and bug fixes must have tests. Documentation may need to
   be updated. If you're unsure what to update, send the PR, and we'll discuss
   in review.
4. Note that PRs updating dependencies and new Go versions are not accepted.
   Please file an issue instead.
5. A changelog entry is required for user facing issues.
-->

**What type of PR is this?**
 Bug fix


**What does this PR do? Why is it needed?**
Fix `getStateRandao` to handle historical queries correctly, returning RANDAO mixes as specified by the Beacon API instead of an error, consistent with clients like Nimbus, Teku, and Lodestar.

I referred to Teku's implementation:
```java
if (epoch.isGreaterThan(stateEpoch)) {
  return Optional.empty();
} else if (epoch.isLessThan(stateEpoch)
    && stateEpoch.minusMinZero(epochsPerHistoricalVector).isGreaterThan(0)
    && stateEpoch.minusMinZero(epochsPerHistoricalVector).isGreaterThanOrEqualTo(epoch)) {
  // ignoring the first period of epoch=0 to epochsPerHistoricalVector,
  // return empty if the epoch is not within `epochsPerHistoricalVector` of the state epoch
  return Optional.empty();
}
```

**Which issues(s) does this PR fix?**

Fixes #15649

**Other notes for review**
- Please let me know if I’ve misunderstood any part of the RANDAO logic.

**Acknowledgements**

- [x] I have read [CONTRIBUTING.md](https://github.com/prysmaticlabs/prysm/blob/develop/CONTRIBUTING.md).
- [x] I have included a uniquely named [changelog fragment file](https://github.com/prysmaticlabs/prysm/blob/develop/CONTRIBUTING.md#maintaining-changelogmd).
- [x] I have added a description to this PR with sufficient context for reviewers to understand this PR.

Muzry committed:
fix getStateRandao not returning historic RANDAO mix values

rkapka commented on beacon-chain/rpc/eth/beacon/handlers_state.go:
```diff
@@ -112,7 +112,7 @@ func (s *Server) GetRandao(w http.ResponseWriter, r *http.Request) {
 	if uint64(stEpoch) > uint64(st.RandaoMixesLength()) {
 		randaoEpochLowerBound = uint64(stEpoch) - uint64(st.RandaoMixesLength())
 	}
-	if epoch > stEpoch || uint64(epoch) < randaoEpochLowerBound+1 {
```
This is incorrect. The lower bound for RANDAO is `[state epoch] - [randao mixes length] + 1`, which means removing `+1` introduces a bug. The `"epoch too old"` test case should fail, but it doesn't because of a bug in tests themselves...

We need to change the code in the beginning of `"head state below EpochsPerHistoricalVector"` to this, otherwise future test cases are affected by changes made here.
```
s := &Server{
			Stater: &testutil.MockStater{
				BeaconState: headSt,
			},
			HeadFetcher:           chainService,
			OptimisticModeFetcher: chainService,
			FinalizationFetcher:   chainService,
			BeaconDB:              db,
		}
``` 

Once you change this, I'm sure your code will fail the `"epoch too old"` test case. The proper fix is to leave this `if` statement as-is and change the previous one to the following:
```
if uint64(stEpoch) >= uint64(st.RandaoMixesLength()) {
		randaoEpochLowerBound = uint64(stEpoch) - uint64(st.RandaoMixesLength()) + 1
	}
```

That way the lower bound will be kept `0` for initial states

Muzry replied:
Thank you for your reply and the suggestion on how to fix this issue. 
I added the code to the head state below `EpochsPerHistoricalVector`, but the unit test failed as you expected. I will update the code accordingly and split the test into two separate tests to improve clarity. I appreciate your guidance, and once the code is updated, please take another look.

rkapka commented on beacon-chain/rpc/eth/beacon/handlers_state_test.go:
```diff
@@ -221,6 +221,40 @@ func TestGetRandao(t *testing.T) {
 		assert.Equal(t, http.StatusBadRequest, e.Code)
 		require.StringContains(t, "Epoch is out of range for the randao mixes of the state", e.Message)
 	})
+	t.Run("early epoch scenario - epoch 0 from epoch 13", func(t *testing.T) {
```
Can we have these two tests instead? Having two makes `if uint64(stEpoch) >= uint64(st.RandaoMixesLength())` fail when changing `>=` to `>`

```
t.Run("early epoch scenario - epoch 0 from state at epoch (EpochsPerHistoricalVector - 1)", func(t *testing.T) {
		// Create a state at early epoch
		earlyEpochState, err := util.NewBeaconState()
		require.NoError(t, err)
		earlyEpoch := params.BeaconConfig().EpochsPerHistoricalVector - 1
		require.NoError(t, earlyEpochState.SetSlot(params.BeaconConfig().SlotsPerEpoch*primitives.Slot(earlyEpoch)))

		// Set up RANDAO mix for epoch 0
		// In real networks, this would be the ETH1 block hash used for genesis
		epoch0Randao := bytesutil.ToBytes32([]byte("epoch0"))
		require.NoError(t, earlyEpochState.UpdateRandaoMixesAtIndex(0, epoch0Randao))

		earlyServer := &Server{
			Stater: &testutil.MockStater{
				BeaconState: earlyEpochState,
			},
			HeadFetcher:           &chainMock.ChainService{},
			OptimisticModeFetcher: &chainMock.ChainService{},
			FinalizationFetcher:   &chainMock.ChainService{},
		}

		// Query epoch 0 from state at epoch (EpochsPerHistoricalVector - 1) - should succeed
		request := httptest.NewRequest(http.MethodGet, "http://example.com//eth/v1/beacon/states/{state_id}/randao?epoch=0", nil)
		request.SetPathValue("state_id", "head")
		writer := httptest.NewRecorder()
		writer.Body = &bytes.Buffer{}

		earlyServer.GetRandao(writer, request)
		require.Equal(t, http.StatusOK, writer.Code, "Early epoch queries should succeed when within bounds")

		resp := &structs.GetRandaoResponse{}
		require.NoError(t, json.Unmarshal(writer.Body.Bytes(), resp))
		assert.Equal(t, hexutil.Encode(epoch0Randao[:]), resp.Data.Randao)
	})
	t.Run("early epoch scenario - epoch 0 from state at epoch EpochsPerHistoricalVector", func(t *testing.T) {
		// Create a state at early epoch
		earlyEpochState, err := util.NewBeaconState()
		require.NoError(t, err)
		earlyEpoch := params.BeaconConfig().EpochsPerHistoricalVector
		require.NoError(t, earlyEpochState.SetSlot(params.BeaconConfig().SlotsPerEpoch*primitives.Slot(earlyEpoch)))

		// Set up RANDAO mix for epoch 0
		// In real networks, this would be the ETH1 block hash used for genesis
		epoch0Randao := bytesutil.ToBytes32([]byte("epoch0"))
		require.NoError(t, earlyEpochState.UpdateRandaoMixesAtIndex(0, epoch0Randao))

		earlyServer := &Server{
			Stater: &testutil.MockStater{
				BeaconState: earlyEpochState,
			},
			HeadFetcher:           &chainMock.ChainService{},
			OptimisticModeFetcher: &chainMock.ChainService{},
			FinalizationFetcher:   &chainMock.ChainService{},
		}

		// Query epoch 0 from state at epoch EpochsPerHistoricalVector - should fail
		request := httptest.NewRequest(http.MethodGet, "http://example.com//eth/v1/beacon/states/{state_id}/randao?epoch=0", nil)
		request.SetPathValue("state_id", "head")
		writer := httptest.NewRecorder()
		writer.Body = &bytes.Buffer{}

		earlyServer.GetRandao(writer, request)
		require.Equal(t, http.StatusBadRequest, writer.Code)
		e := &httputil.DefaultJsonError{}
		require.NoError(t, json.Unmarshal(writer.Body.Bytes(), e))
		assert.Equal(t, http.StatusBadRequest, e.Code)
		require.StringContains(t, "Epoch is out of range for the randao mixes of the state", e.Message)
	})

Muzry committed:
update: the lower bound statement

Muzry commented:
Sorry for the late reply, @rkapka.

When I updated the code to:

```go
if uint64(stEpoch) >= uint64(st.RandaoMixesLength()) {
	randaoEpochLowerBound = uint64(stEpoch) - uint64(st.RandaoMixesLength()) + 1
}
if epoch > stEpoch || (uint64(epoch) < randaoEpochLowerBound+1) {
	httputil.HandleError(w, "Epoch is out of range for the randao mixes of the state", http.StatusBadRequest)
	return
}
```
and ran the unit tests, I got some unexpected failures in:
	•	old epoch requested
	•	early epoch scenario - epoch 0 from state at epoch (EpochsPerHistoricalVector - 1)

After rethinking the lower bound issue, I changed the condition to:
```go
if uint64(stEpoch) >= uint64(st.RandaoMixesLength()) {
	randaoEpochLowerBound = uint64(stEpoch) - uint64(st.RandaoMixesLength()) + 1
}
if epoch > stEpoch || (uint64(epoch) < randaoEpochLowerBound) {
	httputil.HandleError(w, "Epoch is out of range for the randao mixes of the state", http.StatusBadRequest)
	return
}
```
With this change, all unit tests pass. I think the reason is that when stateEpoch is 65535, the valid randaoEpoch range should be [0, 65535] (since the RandaoMixes length is 65536). Epoch 0 should be included in the lower bound, while the previous statement excluded it by making the lower bound start at epoch 1.

I’ve pushed the latest code here — please let me know if I’m still missing something.

Muzry committed:
Merge branch 'develop' into fix-state-randao

Muzry committed:
Merge branch 'develop' into fix-state-randao

Muzry committed:
Merge branch 'develop' into fix-state-randao

Muzry committed:
Merge branch 'develop' into fix-state-randao

Muzry committed:
Merge branch 'develop' into fix-state-randao

Muzry committed:
Merge branch 'develop' into fix-state-randao

Muzry committed:
Merge branch 'develop' into fix-state-randao

Muzry committed:
Merge branch 'develop' into fix-state-randao

rkapka commented on changelog/muzry_fix_state_randao.md:
```diff
@@ -0,0 +1,2 @@
+### Fixed
+Fix getStateRandao not returning historic RANDAO mix values
```
```suggestion

Fix getStateRandao not returning historic RANDAO mix values
```

rkapka committed:
Update changelog/muzry_fix_state_randao.md

rkapka commented:
@Muzry I agree with your reasoning. The PR looks great, I am going to merge it

rkapka commented on changelog/muzry_fix_state_randao.md:
```diff
@@ -0,0 +1,3 @@
+### Fixed
+
+Fix getStateRandao not returning historic RANDAO mix values
```
```suggestion
- Fix getStateRandao not returning historic RANDAO mix values
```

rkapka committed:
Update changelog/muzry_fix_state_randao.md
